tributors/main/orcid.py

```diff
@@ -132,77 +130,119 @@ def get_orcid_token():
+            r["family-names"],
+            r["given-names"],
+            r["orcid-id"],
+        )
+        if r["institution-name"]:
+            record = "%s\n  Institutions: %s" % (
+                record,
+                ", ".join(r["institution-name"]),
+            )
+        if r["other-name"]:
+            record = "%s\n  Other Names: %s" % (record, ", ".join(r["other-name"]))
+        if r["email"]:
+            record = "%s\n  Email: %s" % (record, r["email"])
+        if not interactive:
+            print("%s\n" % record)
+        else:
+            print("[%s]\n%s\n" % (idx, record))
+
+    # If interactive, ask for choice prompt
+    if interactive:
+        choices = [str(i) for i, _ in enumerate(results)] + ["s", "S", "skip"]
+        prefix = "1:%s or s to skip" % ("10" if len(results) > 10 else len(results))
+        choice = choice_prompt(
+            "Please enter a choice, or s to skip.",
+            choices=choices,
+            choice_prefix=prefix,
+            multiple=False,
+        )
 
-def get_orcid(email, token, name=None):
-    """Given a GitHub repository address, retrieve a lookup of contributors
-       from the API endpoint. We look to use the GITHUB_TOKEN if exported
-       to the environment, and exit if the response has any issue
+        if not choice or choice in ["s", "S", "skip"]:
+            return
+
+        # Return the orcid identifier
+        return results[int(choice) - 1]["orcid-id"]
+
+
+def get_orcid(email, name=None, interactive=False):
+    """Get an orcid identifier for a given email or name.
     """
-    # We must have a token, and an email OR name
-    if not token and (not email and not name):
+    # We must have an email OR name
+    if not email and not name:
         return
 
     # First look for records based on email
     orcid_id = None
     if email:
-        url = "https://pub.orcid.org/v3.0/search/?q=email:%s" % email
-        orcid_id = record_search(url, token, email)
+        url = "https://pub.orcid.org/v3.0/expanded-search?q=email:%s" % email
+        orcid_id = record_search(url, email, interactive)
 
     # Attempt # 2 will use the first and last name
     if name is not None and not orcid_id:
 
-        # If no comma, likely a space delimiter
         delim = "," if "," in name else " "
         cleaner = "," if delim == " " else " "
 
         parts = name.split(delim)
+
+        # No go if only a first or last name
+        if len(parts) == 1:
+            bot.debug(f"Skipping {name}, first and last are required for search.")
+            return orcid_id
+
         last, first = parts[0].strip(cleaner), " ".join(parts[1:]).strip(cleaner)
-        url = (
-            "https://pub.orcid.org/v3.0/search/?q=given-names:%s+AND+family-name:%s"
-            % (first, last)
-        )
-        orcid_id = record_search(url, token, name)
+        url = "https://pub.orcid.org/v3.0/expanded-search?q=%s+AND+%s" % (first, last)
```

Comment:
shouldn't `first` and `last` here and everywhere be urlencoded?  I see that there is a test and it passes but "in principle" and to avoid any surprise in the future?

Reply:
if any of them missing/empty somehow -- would it still work? if such a basic query not carrying about first/last -- should it be smth like `'+AND+'.join(parts)`  or if it does make a difference to provide last name as one thing, then I would have
```python
name_parts = filter(bool, (first, last));
url = "%s%s" % (url_query, '+AND+'.join(name_parts))
```
or even better came up with a helper
```python
def get_queryurl(query, parts):
   return "https://pub.orcid.org/v3.0/%s/?q=%s" \
        % (url_query, '+AND+'.join(map(urlencode, filter(bool, parts))))
```

since that `+AND+` concatenation is used in other queries and other things might need to be urlencoded (didn't try/check if needed to yet)

Reply:
A helper function would be good, but I don't see where url_query is coming from, and is there a way to make this more clear to the developer what is going on? I don't know what `filter(bool, parts)` is actually doing, along with the map.

Reply:
sorry, `url_query` is from an earlier comment which I removed but forgot that opening of the comment here. just go with the helper ;)

re `filter`, `map`: good old python elements of functional programming, which was promised to be killed by list comprehensions but they never went away (and never should): random google hit https://book.pythontips.com/en/latest/map_filter.html . Official docs have them in https://docs.python.org/3/library/functions.html

Reply:
We just take one line to define the url to pass to record_search, so I don't think adding another function is going to be super helpful here. But it definitely is more confusing with the functional stuffs.

Reply:
Functional or not, I still wonder if url encode is simply necessary to make it correct

Reply:
It seems to work as is, so I don’t think so.